cmudig/draco2
# refactor: migrate repository to `uv` & simplify Pyodide build

## How to test this PR?

`git checkout` this branch, then execute `make`. Everything should work without any further action (linting, tests, typechecks, doc-gen).

Running all the `make` targets still takes a while, because `mypy` and `pytype` are really slow and generating the Jupyter Book output also requires patience, given that we execute all notebooks on the fly to generate the HTML bundle.

> [!NOTE]  
> Please make sure you have `uv` installed and available on your `PATH`.
> You can refer to the installation instructions here: https://docs.astral.sh/uv/getting-started/installation/

## CHANGELOG

- Drop Poetry in favor of uv
- Move `draco` module source under `src/`
- Use Python 3.12.7
- Update `.github/workflows/test.yml` CI flow
- Make devcontainer `Dockerfile` slimmer (~2x image build performance improvement)
- Define `[project.optional-dependencies]` in `pyproject.toml`, allowing `pip install draco` to pull only core dependencies
  - `pip install 'draco[server]'`: will include FastAPI deps
  - `pip install 'draco[renderer]'`: will include deps to render chart specs
  - `pip install 'draco[learn]'`: will include `scikit-learn` enabling `draco.learn.train_model`
- Stop maintaining custom Pyodide distribution, since `clingo` and `vega-datasets` are now members of the official Pyodide distribution (available through https://cdn.jsdelivr.net/pyodide/<verrsion>/full/)
- Build JupyterLite into docs via [jupyterlite-sphinx](https://pypi.org/project/jupyterlite-sphinx/).

## Pyodide Build Simplifications

The complexity of the current Pyodide build pipeline stems from the fact that `clingo` is an impure Python package (depends on `cffi`) without an official recipe for it in the Pyodide repo. In parallel to this PR I am opening PRs against https://github.com/pyodide/pyodide to include packages we need into their official build. This way, we can dramatically reduce the code we maintain in this repo responsible for WASM builds.

### Foreign PRs

- https://github.com/pyodide/pyodide/pull/5183 
- https://github.com/pyodide/pyodide/pull/5184

## Closes

Closes #895
Closes #884
